File: Py/day-02.py
import re
# The 're' module is imported for powerful, optimized regular expression processing.


# --- Regular Expression Setup ---

# Part 1 Pattern: Check for numbers made of EXACTLY two identical halves (e.g., 1212, 55).
# r'^(\d+)\1$' breakdown:
# ^          : Start of the string.
# (\d+)      : Capture Group 1: Match one or more digits.
# \1         : Match the exact content captured by Group 1.
# $          : End of the string.
# The re.compile() function pre-processes the pattern string into a pattern object,
# which makes repeated searching inside the loop faster.
pattern_a = re.compile(r"^(\d+)\1$")

# Part 2 Pattern: Check for numbers made of a repeating pattern (e.g., 1212, 111, 123123123).
# r'^(\d+)\1+$' breakdown:
# ^          : Start of the string.
# (\d+)      : Capture Group 1: Match one or more digits (the base repeating block).
# \1+        : Match the exact content of Group 1 *one or more* additional times.
# $          : End of the string.
pattern_b = re.compile(r"^(\d+)\1+$")


# --- Initialization ---

# 'a' sums the invalid numbers for the Part 1 rule (exactly two halves).
a = 0
# 'b' sums the invalid numbers for the Part 2 rule (any repeating pattern, which includes Part 1).
b = 0

# --- File Input and Parsing ---

# Read the entire file content into a single string.
file_content = open("input-02-2025.txt").read()

# Use re.findall to extract all range pairs (X-Y) from the input string.
# r'(\d+)-(\d+)' breakdown:
# (\d+)      : Capture Group 1: Matches and captures the starting number (lo).
# -          : Matches the literal dash separator.
# (\d+)      : Capture Group 2: Matches and captures the ending number (hi).
# The result is a list of tuples: [('1061119', '1154492'), ('3', '23'), ...]
# The loop iterates through these tuples.
for lo, hi in re.findall(r"(\d+)-(\d+)", file_content):
    # Convert the string boundaries to integers once per range.
    start = int(lo)
    end = int(hi)

    # --- Inner Loop: Processing Each Number ---

    # Iterate over every integer 'i' in the defined range (inclusive).
    for i in range(start, end + 1):
        # Convert the current integer 'i' to its string representation only once.
        s = str(i)

        # Check Part 1 rule using the pre-compiled pattern_a and re.search.
        # re.search is efficient because it stops immediately upon finding a match.
        if pattern_a.search(s):
            a += i

        # Check Part 2 rule using the pre-compiled pattern_b.
        # This check is independent of Part 1, summing numbers that match *any* repeating pattern.
        if pattern_b.search(s):
            b += i

# --- Output ---
# Print the final sums for both parts of the problem.
print(a, b)

File: JS/day-02.js
"use strict"; // Enables strict mode to catch common coding errors (like undeclared variables)

/**
 * Node.js 'fs' module for File System access.
 * We use this to read the puzzle input from the local directory.
 */
const fs = require("fs");

/**
 * --- REGEX PATTERNS ---
 * Both patterns use "Backreferences" (\1) to detect repetition.
 * * patternA: Matches strings composed of exactly two identical parts.
 * Example: "1212" -> Group 1 is "12", \1 matches "12". Result: Match.
 * * patternB: Matches strings where the first part repeats one OR MORE times.
 * Example: "121212" -> Group 1 is "12", \1+ matches "1212". Result: Match.
 */
const patternA = /^(\d+)\1$/;
const patternB = /^(\d+)\1+$/;

// Accumulators for our two-part solution
let a = 0; // Total sum for Part 1
let b = 0; // Total sum for Part 2

/**
 * File Reading:
 * We read the entire file into memory as a UTF-8 string.
 */
const fileContent = fs.readFileSync("inputs/input-02-2025.txt", "utf8");

/**
 * rangeRegex:
 * Looks for the pattern "Number-Number" (e.g., "100-120").
 * The 'g' flag (global) allows us to find all occurrences in the file.
 */
const rangeRegex = /(\d+)-(\d+)/g;
let match;

/**
 * --- MAIN LOOP ---
 * .exec() returns a match object containing capturing groups:
 * match[0]: The full string (e.g., "100-120")
 * match[1]: The first capture group (the start number)
 * match[2]: The second capture group (the end number)
 */
while ((match = rangeRegex.exec(fileContent)) !== null) {
  // Convert the string matches into base-10 integers
  const start = parseInt(match[1], 10);
  const end = parseInt(match[2], 10);

  /**
   * --- RANGE ITERATION ---
   * We iterate through every integer in the inclusive range.
   */
  for (let i = start; i <= end; i++) {
    // Regex .test() requires a string. We cast the number to a string here.
    const s = String(i);

    // Part 1: Check if the number repeats exactly twice
    if (patternA.test(s)) {
      a += i;
    }

    // Part 2: Check if the number repeats two or more times
    if (patternB.test(s)) {
      b += i;
    }
  }
}

// Output the final sums to the terminal
console.log(a, b);
